### PR TITLE
Earth photos

### PR DESCRIPTION
**Pull request by Kronifer**
## Added .earth_photos command
Closes issue #174 

## Description
Added a .earth_photos command by using the unsplash api

## Reasoning
An approved issue I was assigned to.

## Screenshots
Screenshot_20210303-143312_Discord

## Additional Details
Please add environment variable UNSPLASH_KEY and set it to the key for an unsplash app created on their API
In variable `rf`, in field `utm_source=sir_lancebot`, replace sir_lancebot with the name of your unsplash app

## Did you:
- [x] Join the [**Python Discord Community**](https://discord.gg/python)?
- [x] If dependencies have been added or updated, run `pipenv lock`?
- [x] **Lint your code** (`pipenv run lint`)?
- [x] Set the PR to **allow edits from contributors**?

### CODE REVIEW

**Comment by Kronifer** [comment]
Hi there, I ran the linter but it seems to not have returned the errors that are showing on the server check.

**Comment by Xithrius** [comment]
What command did you run?

**Comment by Kronifer** [comment]
@Xithrius  pipenv run lint

**Comment by Xithrius** [comment]
If you haven't already, try running `pipenv run precommit`, and then `pipenv run lint`.

**Comment by Kronifer** [comment]
I'm re-running it now.

**Comment by Kronifer** [comment]
@Xithrius I ran pre-commit, it's buffering on Flake8 right now

**Comment by Xithrius** [comment]
By installing the precommit hook, committing will be blocked until you fix all of its complaints.

**Comment by Kronifer** [comment]
@Xithrius it didn't complain when run but checks failed on github action. Flake8 passed now.

File: bot/exts/easter/earth_photos.py
Comment (Shivansh-007) [inline bot/exts/easter/earth_photos.py:4]: Since the discord.py bot framework is asynchronous, the requests library cannot be used. You can check out other cogs in this repository for examples on how to use aiohttp to make asynchronous web requests.

File: bot/exts/easter/earth_photos.py
Comment (Shivansh-007) [inline bot/exts/easter/earth_photos.py:17]:
```suggestion
    def __init__(self, bot: commands.Bot) -> None:
```
`init` has no special meaning in Python. It's just a method.

`__init__` is the magic method that's called automatically on object instantiation.

File: bot/exts/easter/earth_photos.py
Comment (Shivansh-007) [inline bot/exts/easter/earth_photos.py:18]: Init functions don’t require a docstring.
```suggestion
```

File: bot/exts/easter/earth_photos.py
Comment (Shivansh-007) [inline bot/exts/easter/earth_photos.py:31]: Can we break this down into lines for easier readability, for example like this:
```py
embed = discord.Embed(
     title=random.choice(ERROR_REPLIES),
     color=Colours.soft_red,
     description="You don't have permission to view this channel."
 )
```

File: bot/exts/easter/earth_photos.py
Comment (Shivansh-007) [inline bot/exts/easter/earth_photos.py:38]:
```suggestion
    """Load the Earth Photos cog."""
```
We use a standard load doc-string for all the cogs.

File: bot/exts/easter/earth_photos.py
Comment (Shivansh-007) [inline bot/exts/easter/earth_photos.py:11]: Please remove this wrapper.

**Review by Shivansh-007 — CHANGES_REQUESTED** [review]
Please consider making these changes.  

Sorry for hitting you with all of this at once.  

Thank you for contributing to Sir-Lancebot!

**Review by HassanAbouelela — CHANGES_REQUESTED** [review]
This PR is looking good so far, but it is not in compliance with the unsplash API, so I am blocking it. The rules it violates are:

[API Guidelines, Technical Guidelines Section, Point 2](https://help.unsplash.com/en/articles/2511245-unsplash-api-guidelines): "you must send a request to the download endpoint". See [here](https://help.unsplash.com/en/articles/2511258-guideline-triggering-a-download) for a more complete review of how this should be done.
[API Guidelines, Technical Guidelines Section, Point 3](https://help.unsplash.com/en/articles/2511245-unsplash-api-guidelines): "..your application must... contain a link back to their Unsplash profile."

Note: all the API requirements are recorded [here](https://help.unsplash.com/en/articles/2511245-unsplash-api-guidelines). I hate to ask this because you already did the work, but is this API really worth using when we have to jump through this many hoops?

**Comment by Kronifer** [comment]
@Xithrius @HassanAbouelela so as I'm seeing it, I need to do the following:

[ ] Switch to aiohttp
[ ] Remove unneeded modules
[ ] Change docstrings
[ ] Reformat embeds
[ ] Make sure stuff complies with the unsplash API

Give me a day or two and I'll try to fix these!

**Comment by Kronifer** [comment]
> This PR is looking good so far, but it is not in compliance with the unsplash API, so I am blocking it. The rules it violates are:
> 
> [API Guidelines, Technical Guidelines Section, Point 2](https://help.unsplash.com/en/articles/2511245-unsplash-api-guidelines): "you must send a request to the download endpoint". See [here](https://help.unsplash.com/en/articles/2511258-guideline-triggering-a-download) for a more complete review of how this should be done.
> [API Guidelines, Technical Guidelines Section, Point 3](https://help.unsplash.com/en/articles/2511245-unsplash-api-guidelines): "..your application must... contain a link back to their Unsplash profile."
> 
> Note: all the API requirements are recorded [here](https://help.unsplash.com/en/articles/2511245-unsplash-api-guidelines). I hate to ask this because you already did the work, but is this API really worth using when we have to jump through this many hoops?

Probably not but I'll do it anyway because I've done a lot of it already

**Comment by Kronifer** [comment]
@HassanAbouelela The issues you presented should be fixed.

File: bot/exts/easter/earth_photos.py
Comment (ChrisLovering) [inline bot/exts/easter/earth_photos.py:21]: Use `self.bot.http_session` here instead of making a new client session

File: bot/exts/easter/earth_photos.py
Comment (ChrisLovering) [inline bot/exts/easter/earth_photos.py:23]: We should pass params using a dict and the `params` kwarg with aiohttp, rather than adding directly to the url.

File: bot/exts/easter/earth_photos.py
Comment (ChrisLovering) [inline bot/exts/easter/earth_photos.py:38]: We should use a colour from `bot.constants.Colours` rather than defining it here.
If there isn't one suitable, add one there.

File: Pipfile
Comment (ChrisLovering) [inline Pipfile:19]: I don't think either of these changes to the Pipfile are needed, could you please revert these changes?

File: bot/exts/easter/earth_photos.py
Comment (ChrisLovering) [inline bot/exts/easter/earth_photos.py:15]: Is this used for anything?

File: bot/exts/easter/earth_photos.py
Comment (ChrisLovering) [inline bot/exts/easter/earth_photos.py:33]: Same as above, use the `params` kwarg here, rather than adding to the url

File: bot/exts/easter/earth_photos.py
Comment (ChrisLovering) [inline bot/exts/easter/earth_photos.py:40]: Could you change this to a f-string?

File: bot/exts/easter/earth_photos.py
Comment (ChrisLovering) [inline bot/exts/easter/earth_photos.py:34]: What is this line supposed to do?

**Review by ChrisLovering — CHANGES_REQUESTED** [review]

File: bot/exts/easter/earth_photos.py
Comment (Kronifer) [reply bot/exts/easter/earth_photos.py:21]: oh, okay!

File: Pipfile
Comment (Kronifer) [reply Pipfile:19]: sure!

File: bot/exts/easter/earth_photos.py
Comment (Kronifer) [reply bot/exts/easter/earth_photos.py:15]: you mean self.current_channel? I got rid of it i thought

File: bot/exts/easter/earth_photos.py
Comment (Kronifer) [reply bot/exts/easter/earth_photos.py:40]: sure

File: bot/exts/easter/earth_photos.py
Comment (Kronifer) [reply bot/exts/easter/earth_photos.py:34]: https://github.com/python-discord/sir-lancebot/pull/609#pullrequestreview-602548202

**Review by Kronifer — COMMENTED** [review]
i added the changes requested

File: bot/exts/easter/earth_photos.py
Comment (ChrisLovering) [reply bot/exts/easter/earth_photos.py:34]: Ah, so this line specifically is doing nothing, we just need to send the get request?

Comment (Kronifer) [reply bot/exts/easter/earth_photos.py:34]: yeah, but i needed to call `er.status` or it wouldn't lint

**Comment by Kronifer** [comment]
@ChrisLovering self.bot.http_session only works once.

**Comment by Kronifer** [comment]
@ChrisLovering never mind, fixed

**Comment by HassanAbouelela** [comment]
I've dismissed my review for now, as it seems we are now in compliance with the API's requirements. I tried to make sure we meet them to the best of my ability, but I am not a lawyer. If someone is more legally inclined, or if someone has time to sort through a bunch of requirements, I'd appreciate a second set of eyes looking over: https://help.unsplash.com/en/articles/2511245-unsplash-api-guidelines.

**Comment by Kronifer** [comment]
@ChrisLovering i have made the changes requested.

**Comment by Kronifer** [comment]
it seems pipenv failed install on the lint check, should I re-commit or will someone re-run it?

**Comment by Kronifer** [comment]
It seems I forgot to lock, fixing now.

**Comment by Kronifer** [comment]
Alright, lint passed!

File: Pipfile.lock
Comment (Shivansh-007) [inline Pipfile.lock:591]: Can you reset this to `upstream/master`.

File: bot/exts/easter/earth_photos.py
Comment (Shivansh-007) [reply bot/exts/easter/earth_photos.py:7]: This has still not being fixed and the wrapper exists.

File: bot/exts/easter/earth_photos.py
Comment (Shivansh-007) [inline bot/exts/easter/earth_photos.py:38]:
```suggestion
                color=Colours.grass_green
            )
```

File: bot/exts/easter/earth_photos.py
Comment (Shivansh-007) [inline bot/exts/easter/earth_photos.py:34]: Uhhh what does line do?

File: bot/exts/easter/earth_photos.py
Comment (Shivansh-007) [inline bot/exts/easter/earth_photos.py:33]:
```suggestion
                    params={"client_id": UnClient_id}
            ) as er:
```

File: bot/exts/easter/earth_photos.py
Comment (Shivansh-007) [inline bot/exts/easter/earth_photos.py:29]: Can you remove the redundant variables here? Like the `userlinks `.

File: bot/exts/easter/earth_photos.py
Comment (Shivansh-007) [inline bot/exts/easter/earth_photos.py:22]:
```suggestion
                    params={"query": "earth", "client_id": UnClient_id}
            ) as r:
```

**Review by Shivansh-007 — CHANGES_REQUESTED** [review]
I like where this PR is going. Please consider making these changes.

File: Pipfile.lock
Comment (Kronifer) [reply Pipfile.lock:591]: I didn't know that happened, sure

File: bot/exts/easter/earth_photos.py
Comment (Kronifer) [reply bot/exts/easter/earth_photos.py:34]: https://github.com/python-discord/sir-lancebot/pull/609#pullrequestreview-602548202

File: bot/exts/easter/earth_photos.py
Comment (Kronifer) [reply bot/exts/easter/earth_photos.py:30]: I have to keep them for getting specific data, unless there's a better way, I don't work with web requests much

File: bot/exts/easter/earth_photos.py
Comment (Shivansh-007) [reply bot/exts/easter/earth_photos.py:36]: Sorry I didn’t get you.

File: bot/exts/easter/earth_photos.py
Comment (Shivansh-007) [reply bot/exts/easter/earth_photos.py:30]: `userlinks ` is being used anywhere and even a few others so I don’t see a need of them.

File: bot/exts/easter/earth_photos.py
Comment (Kronifer) [reply bot/exts/easter/earth_photos.py:36]: https://help.unsplash.com/en/articles/2511245-unsplash-api-guidelines

File: bot/exts/easter/earth_photos.py
Comment (Kronifer) [reply bot/exts/easter/earth_photos.py:30]: userlinks is used to get profile link for embed, downloadlinksdata for the download links, linksdata for the embed links, embedlink is self-explanitory, userdata is also self explanitory, username is for crediting the person who made the image as of [here](https://help.unsplash.com/en/articles/2511245-unsplash-api-guidelines), name crediting is needed for production app, profile is used for a link to the profile

File: Pipfile.lock
Comment (Kronifer) [reply Pipfile.lock:591]: fixed as of https://github.com/python-discord/sir-lancebot/pull/609/commits/825080e805787afc1f36a80b4fa3f73040ebdc75

Comment (Kronifer) [reply Pipfile.lock:591]: Quoting https://github.com/Kronifer/sir-lancebot/blob/earth_photos/Pipfile.lock#L591

**Comment by Kronifer** [comment]
@Shivansh-007 I've fixed everything over in my repo

File: Pipfile.lock
Comment (Shivansh-007) [reply Pipfile.lock:591]: I am, and it is still not solved.

File: bot/exts/easter/earth_photos.py
Comment (Shivansh-007) [reply bot/exts/easter/earth_photos.py:36]: Am talking about `er.status`.

File: bot/exts/easter/earth_photos.py
Comment (Shivansh-007) [reply bot/exts/easter/earth_photos.py:7]: This is still not fixed.

File: bot/exts/easter/earth_photos.py
Comment (Kronifer) [reply bot/exts/easter/earth_photos.py:36]: I have to call `er.status` otherwise the lint fails

Comment (Kronifer) [reply bot/exts/easter/earth_photos.py:36]: because then flake8 returns `er is defined but never used`

File: Pipfile.lock
Comment (Kronifer) [reply Pipfile.lock:591]: Oh wow, Im stupid! Sorry @Shivansh-007 , I looked at it wrong

**Comment by Kronifer** [comment]
@Shivansh-007 it seems some of this is outdated, as the logging thing no longer exists as shown [here](https://github.com/Kronifer/sir-lancebot/blob/earth_photos/bot/exts/easter/earth_photos.py#L7-L9)

File: Pipfile.lock
Comment (Kronifer) [reply Pipfile.lock]: Hi @Shivansh-007 , I cant do this as it breaks pipenv and the packages, sorry!

File: bot/exts/easter/earth_photos.py
Comment (Shivansh-007) [reply bot/exts/easter/earth_photos.py:36]: Flake8 never fails on this issue. You have added it in b2ee1b9e79766b8cfe7b45670f6c86e7166ee84f commit and I don’t see it failing before. And you can replace it with `_`, it is used to indicate that the  result is being deliberately ignored.

File: Pipfile.lock
Comment (Shivansh-007) [reply Pipfile.lock]: It doesn’t beak them. You are doing `pipenv lock`, am asking you to reset it to what it is at  the main branch master that is, what it is [here](https://github.com/python-discord/sir-lancebot/blob/master/Pipfile.lock)

**Comment by Shivansh-007** [comment]
> @Shivansh-007 it seems some of this is outdated, as the logging thing no longer exists as shown [here](https://github.com/Kronifer/sir-lancebot/blob/earth_photos/bot/exts/easter/earth_photos.py#L7-L9)

Which logging thing are you talking about?

File: Pipfile.lock
Comment (Kronifer) [reply Pipfile.lock:591]: sorry, im not sure how to do this, what would i use?

File: bot/exts/easter/earth_photos.py
Comment (Kronifer) [reply bot/exts/easter/earth_photos.py:35]: oh sorry, i ran flake8 locally and it freaked on me

File: Pipfile.lock
Comment (Shivansh-007) [reply Pipfile.lock:591]: Copy the `Pipfile.lock` from python-discord/sir-lancebot and paste it into yours.

**Comment by Kronifer** [comment]
> > @Shivansh-007 it seems some of this is outdated, as the logging thing no longer exists as shown [here](https://github.com/Kronifer/sir-lancebot/blob/earth_photos/bot/exts/easter/earth_photos.py#L7-L9)
> 
> Which logging thing are you talking about?

You said something about it in the change requested directly below [here](https://github.com/python-discord/sir-lancebot/pull/609#pullrequestreview-603455089)

**Comment by Kronifer** [comment]
sorry, two under the linked comment

**Comment by Shivansh-007** [comment]
> > > @Shivansh-007 it seems some of this is outdated, as the logging thing no longer exists as shown [here](https://github.com/Kronifer/sir-lancebot/blob/earth_photos/bot/exts/easter/earth_photos.py#L7-L9)
> > 
> > 
> > Which logging thing are you talking about?
> 
> You said something about it in the change requested directly below [here](https://github.com/python-discord/sir-lancebot/pull/609#pullrequestreview-603455089)

I never said anything about logging, I asked you to remove the wrapper I.e. `UnClient_id = Tokens.unsplash_key ` and use `Tokens.unsplash_key` directly.

**Comment by Kronifer** [comment]
oh im dumb

File: bot/exts/easter/earth_photos.py
Comment (Kronifer) [reply bot/exts/easter/earth_photos.py:35]: how do you mean?

File: Pipfile.lock
Comment (Kronifer) [reply Pipfile.lock:591]: ok

Comment (Kronifer) [reply Pipfile.lock:591]: @Shivansh-007 I did this but as soon as I ran pipenv sync --dev locally it reset this

Comment (Kronifer) [reply Pipfile.lock:591]: it failed identify install when I made the change

**Comment by Kronifer** [comment]
@Shivansh-007 all changes have been made

File: bot/exts/easter/earth_photos.py
Comment (ChrisLovering) [inline bot/exts/easter/earth_photos.py:46]: Could you add a check here, so that if the token isn't set the cog doesn't try to load. As it raises an error currently if the token isn't present.
For example from the games cog:
https://github.com/python-discord/sir-lancebot/blob/a3bd208580ff223cd0120d0c4b5a1bf3c28d03e5/bot/exts/evergreen/game.py#L472-L481

File: bot/exts/easter/earth_photos.py
Comment (ChrisLovering) [inline bot/exts/easter/earth_photos.py:7]: Could you please remove this wrapper and use `Tokens.unsplash_key` directly in your param's dict.

**Review by ChrisLovering — CHANGES_REQUESTED** [review]

File: bot/exts/easter/earth_photos.py
Comment (Kronifer) [reply bot/exts/easter/earth_photos.py:7]: I've done this, i just dont have it commited yet as far as I know

File: bot/exts/easter/earth_photos.py
Comment (Kronifer) [reply bot/exts/easter/earth_photos.py:46]: sure!

File: bot/constants.py
Comment (Kronifer) [reply bot/constants.py:275]: this makes sense, okay!

**Comment by Kronifer** [comment]
@ChrisLovering made the requested changes

File: bot/exts/easter/earth_photos.py
Comment (Shivansh-007) [inline bot/exts/easter/earth_photos.py:37]: Can we add a comment here saying why are we doing this?

**Review by ChrisLovering — APPROVED** [review]
Looks good to me!

**Note to core devs**: We should make sure this env var exists in the bot env before mreging, and we should update the [contributing guide](https://pythondiscord.com/pages/contributing/sir-lancebot/sir-lancebot-env-var-reference/) too

**Comment by Kronifer** [comment]
Thanks @ChrisLovering !

File: bot/exts/easter/earth_photos.py
Comment (Shivansh-007) [inline bot/exts/easter/earth_photos.py:46]: Take this inside the indent. As it is part of 
```py
        await with ctx.typing()
```

File: bot/exts/easter/earth_photos.py
Comment (Shivansh-007) [inline bot/exts/easter/earth_photos.py:45]: I don’t like the word 'made' here, the photos are not made but rather added or clicked.

File: bot/exts/easter/earth_photos.py
Comment (Shivansh-007) [inline bot/exts/easter/earth_photos.py:41]:
```suggestion
                description="A photo of Earth 🌎  from Unsplash.",
```
Emoji can be ignored, came up on my keyboard and just put it. It looks good tho

File: bot/exts/easter/earth_photos.py
Comment (Shivansh-007) [inline bot/exts/easter/earth_photos.py:24]:
```suggestion
                    params={"query": "earth", "client_id": Tokens.unsplash_access_key}
            ) as r:
```

**Review by Shivansh-007 — CHANGES_REQUESTED** [review]
Looks awesome! Just a few nitpicks 😛

File: bot/exts/easter/earth_photos.py
Comment (Kronifer) [reply bot/exts/easter/earth_photos.py]: changed to `photo by`, as per the unsplash api production checklist

**Comment by Kronifer** [comment]
@Shivansh-007 all changes made

File: bot/exts/easter/earth_photos.py
Comment (Shivansh-007) [reply bot/exts/easter/earth_photos.py:38]: This hasn’t being mentioned yet.

Comment (Kronifer) [reply bot/exts/easter/earth_photos.py:38]: oh sorry, i'll fix that!

Comment (Kronifer) [reply bot/exts/easter/earth_photos.py:39]: fixed as of commit https://github.com/python-discord/sir-lancebot/pull/609/commits/68ff4f3a524e90099e6953bd601321623e789e66

**Review by Shivansh-007 — APPROVED** [review]
Thanks! Looks Good To Me!

**Comment by Kronifer** [comment]
> Thanks! Looks Good To Me!

No problem!

File: bot/exts/easter/earth_photos.py
Comment (HassanAbouelela) [inline bot/exts/easter/earth_photos.py:47]: There is still a bit more that needs to be added. [This article](https://help.unsplash.com/en/articles/2511315-guideline-attribution) recommends that you have both the author and the site links. It also requires you use a referral link instead of a regular link. Check the article for all the specifics.

Comment (Kronifer) [reply bot/exts/easter/earth_photos.py:47]: oh, whoops! I'll get going on that

**Comment by Kronifer** [comment]
Made a quick update to the info at the top, make sure to take a look as it is important for api verification

**Review by HassanAbouelela — APPROVED** [review]
Very clean PR, and works as expected. We are now blocked by the API application, but the code is ready.

**Comment by Kronifer** [comment]
**to the people merging this:**
It should take 5-10 business days for unsplash to review an application

**Comment by HassanAbouelela** [comment]
Storing these here for the application, but I have the raw versions too.
image
earth command

**Comment by Kronifer** [comment]
awesome! thanks for the animations

**Comment by HassanAbouelela** [comment]
@Kronifer The token is ready, please update the app name to `Sir Lancebot` (I believe in code that would be `Sir%20Lancebot`

**Comment by Kronifer** [comment]
@HassanAbouelela thats done!

**Comment by Kronifer** [comment]
awesome!

**Pull request merged**
